lib/utils/index.js

```diff
@@ -1,11 +1,34 @@
 const crypto = require('crypto');
 
-// todo: generate proper licence numbers
-const generateLicenceNumber = () => {
+const generateLicenceNumber = (model, transaction, type, version) => {
   const buf = crypto.randomBytes(48).toString('hex');
-  const chars = buf.replace(/[0-9]/g, '').substring(0, 2).toUpperCase();
-  const digits = buf.replace(/[a-z]/g, '').substring(0, 6);
-  return `${chars}-${digits}`;
+  const digits = buf.replace(/[a-z]/g, '').substring(0, 8);
+  let licenceNumber;
+  switch (type) {
+    case 'pil':
+      licenceNumber = `I${digits}`;
+      break;
+    case 'project':
+      if (version > 0) {
+        licenceNumber = `PP${digits.substring(0, 7)}`;
+      } else {
+        licenceNumber = `P${digits}`;
+      }
+      break;
+    case 'pel':
+      licenceNumber = `X${digits}`;
+      break;
+  }
+
+  return model
+    .query(transaction)
+    .findOne({ licenceNumber })
+    .then(existing => {
+      if (existing) {
+        generateLicenceNumber(model, transaction, type, version);
```

Comment:
Don't you need to return this here?

Reply:
@wheelsandcogs why ? It should ultimately return licenceNumber ?

Reply:
because you're calling it recursively, and licence number is defined inside the function. So when you exit the call to `generateLicenceNumber(model, transaction, type, version)` you either need to assign the result to `licenceNumber` or just return it directly, otherwise you'll just return the first licence number generated and not the 2nd attempt.

Reply:
Good catch @wheelsandcogs !